src/memory.rs

```diff
@@ -1,48 +1,109 @@
-    pub static MEMORY_MANAGER: RefCell<MemoryManager<VectorMemory>> =
+    static MEMORY_MANAGER: RefCell<MemoryManager<VectorMemory>> =
         RefCell::new(MemoryManager::init(VectorMemory::new(RefCell::new(vec![]))));
     #[cfg(target_arch = "wasm32")]
-    pub static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> =
+    static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> =
         RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
-    pub static METADATA: RefCell<Cell<Metadata, Memory>> = RefCell::new(Cell::init(
-            MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(0))),
-            Metadata::default()).unwrap());
-    pub static AUTH: RefCell<StableBTreeMap<PrincipalStorable, AuthSet, Memory>> = RefCell::new(
-        StableBTreeMap::init(MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(1)))));
-    pub static PROVIDERS: RefCell<StableBTreeMap<u64, Provider, Memory>> = RefCell::new(
-        StableBTreeMap::init(MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(2)))));
-    pub static SERVICE_PROVIDER_MAP: RefCell<StableBTreeMap<StorableRpcService, u64, Memory>> = RefCell::new(
-        StableBTreeMap::init(MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(3)))));
+    static API_KEY_MAP: RefCell<StableBTreeMap<ProviderId, ApiKey, Memory>> =
+        RefCell::new(StableBTreeMap::init(MEMORY_MANAGER.with_borrow(|m| m.get(MemoryId::new(5)))));
+    static MANAGE_API_KEYS: RefCell<ic_stable_structures::Vec<PrincipalStorable, Memory>> =
+        RefCell::new(ic_stable_structures::Vec::init(MEMORY_MANAGER.with_borrow(|m|m.get(MemoryId::new(6)))).expect("Unable to read API key principals from stable memory"));
 }
 
-pub fn get_nodes_in_subnet() -> u32 {
-    UNSTABLE_SUBNET_SIZE.with_borrow(|n| *n)
+pub fn get_api_key(provider_id: ProviderId) -> Option<ApiKey> {
+    API_KEY_MAP.with_borrow_mut(|map| map.get(&provider_id))
 }
 
-pub fn set_nodes_in_subnet(nodes_in_subnet: u32) {
-    UNSTABLE_SUBNET_SIZE.with_borrow_mut(|n| *n = nodes_in_subnet)
+pub fn insert_api_key(provider_id: ProviderId, api_key: ApiKey) {
+    API_KEY_MAP.with_borrow_mut(|map| map.insert(provider_id, api_key));
+}
+
+pub fn remove_api_key(provider_id: ProviderId) {
+    API_KEY_MAP.with_borrow_mut(|map| map.remove(&provider_id));
+}
+
+pub fn is_api_key_principal(principal: &Principal) -> bool {
+    MANAGE_API_KEYS.with_borrow_mut(|principals| {
+        principals
+            .iter()
+            .any(|PrincipalStorable(other)| &other == principal)
+    })
+}
+
+pub fn set_api_key_principals(new_principals: Vec<Principal>) {
+    MANAGE_API_KEYS.with_borrow_mut(|principals| {
+        while !principals.is_empty() {
+            principals.pop();
+        }
+        for principal in new_principals {
+            principals
+                .push(&PrincipalStorable(principal))
+                .expect("Error while adding API key principal");
+        }
+    });
+}
+
+pub fn get_demo_status() -> bool {
```

Comment:
Thanks; yes, I originally had a `DemoStatus` enum but later converted it to a `bool` and forgot to change the name of the function. Updated to `is_demo_active` and `set_demo_active` as a slight adaptation of your suggestion.

Reply:
I thought I responded to this but it might've been lost when refreshing the page. If that was the case, I've updated to `is_demo_active` and `set_demo_active` (originally had an enum called `DemoStatus`).